internal/consensus/state_test.go

```diff
@@ -1609,6 +1609,104 @@ func TestStateLock_POLSafety2(t *testing.T) {
+
+	partSize := types.BlockPartSizeBytes
+
+	proposalCh := subscribe(cs1.eventBus, types.EventQueryCompleteProposal)
+	timeoutWaitCh := subscribe(cs1.eventBus, types.EventQueryTimeoutWait)
+	newRoundCh := subscribe(cs1.eventBus, types.EventQueryNewRound)
+	pv1, err := cs1.privValidator.GetPubKey()
+	require.NoError(t, err)
+	addr := pv1.Address()
+	voteCh := subscribeToVoter(cs1, addr)
+
+	// the block for R0: gets polkad but we miss it
+	// (even though we signed it, shhh)
+	_, propBlock0 := decideProposal(ctx, t, cs1, vss[0], height, round)
+	propBlockHash0 := propBlock0.Hash()
+	propBlockParts0, err := propBlock0.MakePartSet(partSize)
+	require.NoError(t, err)
+	propBlockID0 := types.BlockID{Hash: propBlockHash0, PartSetHeader: propBlockParts0.Header()}
+
+	// the others sign a polka but we don't see it
+	prevotes := signVotes(types.PrevoteType, chainID, propBlockID0, false, vs2, vs3, vs4)
+
+	// the block for round 1
+	prop1, propBlock1 := decideProposal(ctx, t, cs1, vs2, vs2.Height, vs2.Round+1)
+	propBlockParts1, err := propBlock1.MakePartSet(partSize)
+	require.NoError(t, err)
+	propBlockID1 := types.BlockID{Hash: propBlock1.Hash(), PartSetHeader: propBlockParts1.Header()}
+
+	incrementRound(vs2, vs3, vs4)
+
+	round++ // moving to the next round
+	t.Log("### ONTO Round 1")
+	// jump in at round 1
+	startTestRound(cs1, height, round)
+	ensureNewRound(newRoundCh, height, round)
+
+	err = cs1.SetProposalAndBlock(prop1, propBlock1, propBlockParts1, "some peer")
+	require.NoError(t, err)
+	ensureNewProposal(proposalCh, height, round)
+
+	ensurePrevote(voteCh, height, round)
+	validatePrevote(t, cs1, round, vss[0], propBlockID1.Hash)
+
+	signAddVotes(cs1, types.PrevoteType, chainID, propBlockID1, false, vs2, vs3, vs4)
+
+	ensurePrecommit(voteCh, height, round)
+	// the proposed block should now be locked and our precommit added
+	validatePrecommit(t, cs1, round, round, vss[0], propBlockID1.Hash, propBlockID1.Hash)
+
+	// add precommits from the rest
+	signAddVotes(cs1, types.PrecommitType, chainID, types.BlockID{}, true, vs2, vs4)
+	signAddVotes(cs1, types.PrecommitType, chainID, propBlockID1, true, vs3)
+
+	incrementRound(vs2, vs3, vs4)
+
+	// timeout of precommit wait to new round
+	ensureNewTimeout(timeoutWaitCh, height, round, cs1.config.Precommit(round).Nanoseconds())
+
+	round++ // moving to the next round
+	// in round 2 we see the polkad block from round 0
+	newProp := types.NewProposal(height, round, 0, propBlockID0, propBlock0.Header.Time)
+	p := newProp.ToProto()
+	err = vs3.SignProposal(chainID, p)
+	require.NoError(t, err)
+
+	newProp.Signature = p.Signature
+
+	err = cs1.SetProposalAndBlock(newProp, propBlock0, propBlockParts0, "some peer")
+	require.NoError(t, err)
+
+	// Add the pol votes
+	addVotes(cs1, prevotes...)
+
+	ensureNewRound(newRoundCh, height, round)
+	t.Log("### ONTO Round 2")
+	/*Round2
+	// now we see the polka from round 1, but we shouldn't unlock
+	*/
+	ensureNewProposal(proposalCh, height, round)
+
```

Comment:
For when is enabled, the test compares with nil, two lines below.

Reply:
Let me see if I understood this change: you duplicated `TestStateLock_POLSafety2` into two test cases: one with PBTS and one without.
If that's the case, could it be possible to refactor it into two test cases (with and without) but reusing most of the logic of the original test case? It would be much easier to maintain IMO

Reply:
Refactored and merged.